modules/sirdataRtdProvider.js

```diff
@@ -65,162 +63,473 @@ const partnerIds = {
+ * @param {Object} userConsent - The user consent information
+ */
+export function getSegmentsAndCategories(reqBidsConfigObj, onDone, moduleConfig, userConsent) {
+  logInfo(LOG_PREFIX, 'get Segments And Categories');
+  const adUnits = (reqBidsConfigObj && reqBidsConfigObj.adUnits) || getGlobal().adUnits;
+  if (!adUnits) {
+    logInfo(LOG_PREFIX, 'no ad unit, RTD processing is useless');
+    onDone();
+    return;
+  }
+
+  const gdprApplies = deepAccess(userConsent, 'gdpr.gdprApplies') ? userConsent.gdpr.gdprApplies : false;
+  const sirdataSubDomain = params.preprod ? 'kvt-preprod' : 'kvt';
+
+  let euids; // empty if no right to access device (publisher or user reject)
+  let privacySignals = '';
+
+  // Default global endpoint is cookie-based only if no rules falls into cookieless or consent has been given or GDPR doesn't apply
+  if (hasDeviceAccess() && !userConsent.coppa && (isEmpty(userConsent.usp) || userConsent.usp === -1 || (userConsent.usp[0] === '1' && (userConsent.usp[1] !== 'N' && userConsent.usp[2] !== 'Y'))) && (!gdprApplies || (deepAccess(userConsent, 'gdpr.vendorData.vendor.consents') && userConsent.gdpr.vendorData.vendor.consents[GVLID] && deepAccess(userConsent, 'gdpr.vendorData.purpose.consents') && userConsent.gdpr.vendorData.purpose.consents[1] && (userConsent.gdpr.vendorData.purpose.consents[2] || userConsent.gdpr.vendorData.purpose.consents[3]) && userConsent.gdpr.vendorData.purpose.consents[4])) && (isEmpty(userConsent.gpp) || userConsent.gpp.gppString) && checkCookieSupport()) {
+    params.sirdataDomain = 'sddan.com'; // cookie based domain
+    params.cookieAccessGranted = true; // cookies sent in request
+
+    if (gdprApplies && deepAccess(userConsent, 'gdpr.consentString')) {
+      privacySignals = `&gdpr=${gdprApplies}&gdpr_consent=${userConsent.gdpr.consentString}`;
+    } else if (!isEmpty(userConsent.usp)) {
+      privacySignals = `&ccpa_consent=${userConsent.usp.toString()}`;
+    } else if (deepAccess(userConsent, 'gpp.gppString')) {
+      const sid = deepAccess(userConsent, 'gpp.applicableSections') ? `&gpp_sid=${userConsent.gpp.applicableSections.join(',')}` : '';
+      privacySignals = `&gpp=${userConsent.gpp.gppString}${sid}`;
+    }
+
+    // EUIDS from storage and sync global for graph
+    euids = getUidFromStorage(); // Sirdata Id
+    if (!params.avoidPostEids && typeof getGlobal().getUserIdsAsEids === 'function') {
+      euids = mergeEuidsArrays(euids, getGlobal().getUserIdsAsEids()); // for graph id
+    }
+  }
+
+  const url = `https://${sirdataSubDomain}.${params.sirdataDomain}/api/v1/public/p/${params.partnerId.toString()}/d/${params.key.toString()}/s?callback=&allowed_post_content=${!params.avoidPostContent}${privacySignals}${params.actualUrl ? `&url=${encodeURIComponent(params.actualUrl)}` : ''}`;
+  const method = isEmpty(euids) ? 'GET' : 'POST';
+  const payload = isEmpty(euids) ? null : JSON.stringify({ external_ids: euids });
```

Comment:
Generally, publishers will be violating their contracts with eid providers to send this to you to store in the graph. most of them have a 'you will not store' and 'permissable use' portions of their contract. Installing your module would cause publishers to breach many of their contracts. I'm fine if you take a whitelist of eids sources to store in configuration instead of just defaulting to all. Then publishers would be knowingly opting into this for specific id providers.

Reply:
Thank you very much for this clarification :) There is perhaps an important information to bring here, which is that we are talking about graph resolution (in terms of access to a catalog of segments relating to a user), but not about a graph constitution. The majority of these IDs will involve agreements with the ID providers we talk to (already like SSP/DSPs), as most need to be decrypted or they will be useless. This makes the use of a whitelist system almost impossible... But once again the publisher has a control parameter at its disposal, and I think we can consider a blacklist: which id provider should systematically be blocked according to you? Finally, unless I'm mistaken, this use case already exists in the Audigent RTD module (https://github.com/prebid/Prebid.js/blob/master/modules/hadronRtdProvider.js)

Reply:
whoa good catch!

Reply:
I am not sure I follow why a publisher configured whitelist is impossible? Perhaps could you restate?

Reply:
Sure :) In the user ids module the editor only activates the ids that are useful to him and with whom he works. Among these, Sirdata also works (like the bidders) with some, and will later work with others. a whitelist system implies that if Sirdata enters into an agreement with an ID supplier to acquire the decryption keys, we must have all the configs of all the partners updated, which is "impossible" in the short term. At the same time, I think there are two possibilities: either the id is universal and no contractual limitations exist, or the id is private and the user cannot share it, in which case it is encrypted and without agreement with the supplier it cannot be deciphered. In addition, since the encryption keys are changed regularly, the encrypted ID cannot be used for tracking. So there should be no individual limitation, and if by chance there one day exists an ID that absolutely says it is blocked, it is entirely manageable by a blacklist. That being said, we planned that the editor can block absolutely everything with the "avoidPostEids" parameter, but perhaps your recommendation is that it can block one or a few ids and not all?

Reply:
I know of many examples of ids that do not fall within this dichotomy: "either the id is universal and no contractual limitations exist, or the id is private and the user cannot share it, in which case it is encrypted and without agreement with the supplier it cannot be deciphered."

>  we must have all the configs of all the partners updated, which is "impossible" in the short term.

I am afraid I find the risk to the publisher outweighs this concern; circulating within the community for additional feedback. We're also taking a look at Audigent's decision here. If you start working with a new id provider, I would imagine all the publisher configs should be possible to update within just a few months.

Reply:
sourced from the internet: "The Federal Trade Commission (FTC) has taken enforcement actions against companies that have collected, used, or retained personal information for purposes other than the original reason given. The FTC can establish a data minimization rule that considers it an unfair trade practice to collect, use, transfer, or retain personal data beyond what is reasonably necessary. 

Data minimization is a privacy principle that states that personal information should only be collected and retained for as long as necessary to fulfill a specific purpose. The FTC has brought enforcement actions against companies that have:
Collected more personal information than they disclose or need
Used or shared personal information for incompatible purposes
Retained information beyond the period for which the data is required"

Reply:
Thank you for your valuable feedback and for highlighting the privacy risks associated with handling encrypted identifiers (eids). We share your concerns and have carefully considered your recommendations.

In response, we have adjusted our code to provide publishers with control over which eids they share with us via a whitelist system. By default, they will share the ID5 identifier with us, as we intend to run tests that involve ID5 providing us with decryption keys. Additionally, they will share their own identifiers (pubProvidedId and sharedId).

If publishers wish to share other eids with us, such as the zeotap ID or the ramp ID, they can do so after obtaining mutual agreement by using the parameter authorizedEids. Publishers also have the option to share no eids at all by setting authorizedEids to an empty array ([]).

We believe this approach balances the need for data utility with the essential requirement for privacy protection and compliance with contractual obligations.

Thank you once again for your insightful feedback.

Reply:
Thanks!